feat: Add PostgreSQL as alternative database backend

dims opened the pull request:
## Summary

  - Add PostgreSQL as a first-class database backend alongside MongoDB
  - Introduce database abstraction layer enabling seamless provider switching
  - Add owner references to RebootNode CRs for automatic garbage collection
  - Add XID burst detector for pattern analysis in health-events-analyzer

  ## Motivation

  This PR enables NVSentinel to use PostgreSQL as an alternative to MongoDB, providing:
  - **Flexibility**: Choose the database that best fits your infrastructure
  - **Familiarity**: PostgreSQL is widely used and well-understood by many teams
  - **Cost optimization**: Leverage existing PostgreSQL infrastructure if available

  ## Changes

  ### Database Abstraction Layer (Commits 1-4)
  Pure refactoring commits that introduce abstractions without changing MongoDB behavior:
  - `PipelineBuilder` interface for database-agnostic change stream pipelines
  - `QueryBuilder` and `UpdateBuilder` for portable query construction
  - Enhanced datastore interfaces with `Unwrap()` for backward compatibility
  - Updated FQ/ND/FR initializers to use the new abstractions

  ### MongoDB Improvements (Commit 5)
  - **Owner references on RebootNode CRs**: Enables Kubernetes garbage collection to automatically clean up CRs when nodes are deleted

  ### New Features (Commit 6)
  - **XID burst detector**: Detects patterns of repeated XID errors in health-events-analyzer (unrelated to PostgreSQL)

  ### PostgreSQL Implementation (Commits 7-10)
  - Vendored Bitnami PostgreSQL Helm chart for easy deployment
  - Full PostgreSQL client implementing the same interfaces as MongoDB
  - Change stream emulation using PostgreSQL LISTEN/NOTIFY triggers
  - Pipeline filter translation from MongoDB format to SQL
  - Helm templates and values for PostgreSQL deployment
  - CI/CD jobs for PostgreSQL E2E testing
  - Documentation and schema definitions

<!-- Brief description of your changes -->

## Type of Change
- [ ] 🐛 Bug fix
- [x] ✨ New feature
- [ ] 💥 Breaking change
- [x] 📚 Documentation
- [ ] 🔧 Refactoring
- [x] 🔨 Build/CI

## Component(s) Affected
- [x] Core Services
- [x] Documentation/CI
- [x] Fault Management
- [x] Health Monitors
- [x] Janitor
- [ ] Other: ____________

## Testing
- [x] Tests pass locally
- [x] Manual testing completed
- [x] No breaking changes (or documented)

## Checklist
- [ ] Self-review completed
- [ ] Documentation updated (if needed)
- [ ] Ready for review

dims committed:
refactor: Add PipelineBuilder interface for database-agnostic pipelines

Introduce a PipelineBuilder interface that abstracts change stream pipeline
construction. This allows different database providers to implement optimized
pipelines for their specific change stream behavior.

This commit adds:
- PipelineBuilder interface with methods for common pipeline patterns
- MongoDBPipelineBuilder implementing the interface for MongoDB change streams
- Unit tests for pipeline builder functionality

This is a pure refactoring commit - MongoDB behavior is unchanged.

Signed-off-by: Davanum Srinivas <dsrinivas@nvidia.com>

dims committed:
refactor: Add datastore abstraction interfaces and MongoDB adapter enhancements

Enhance the datastore abstraction layer with:
- QueryBuilder and UpdateBuilder interfaces for database-agnostic queries
- FindHealthEventsByQuery and UpdateHealthEventsByQuery methods
- MongoDB adapter enhancements: GetDatabaseClient, GetCollectionClient, Unwrap
- NewChangeStreamWatcher method for config map-based watcher creation
- Watcher factory for creating change stream watchers across providers
- Improved health store with proper RawEvent population and normalization

This enables services to use the datastore abstraction while still having
access to provider-specific functionality when needed.

Signed-off-by: Davanum Srinivas <dsrinivas@nvidia.com>

dims committed:
feat: Add query and update builder for database-agnostic operations

Implement a fluent query builder that generates both MongoDB filters and
PostgreSQL WHERE clauses from the same API. This enables services to write
database-agnostic code that works with both MongoDB and PostgreSQL.

Features:
- Query builder with comparison operators (Eq, Ne, Gt, Gte, Lt, Lte)
- Logical operators (And, Or)
- Collection operators (In, Nin)
- Nested field support with dot notation
- Update builder with Set, Inc, Push, Pull, Unset operations
- Proper JSONB handling for PostgreSQL complex types
- Comprehensive test coverage for both MongoDB and PostgreSQL outputs

Signed-off-by: Davanum Srinivas <dsrinivas@nvidia.com>

dims committed:
refactor: Migrate FQ/ND/FR initializers to use PipelineBuilder

Update fault-quarantine, node-drainer, and fault-remediation to use the
new PipelineBuilder interface instead of calling pipeline construction
functions directly. This enables database-agnostic pipeline creation
while maintaining backward compatibility with MongoDB.

Changes:
- fault-quarantine: Use GetPipelineBuilder().BuildAllHealthEventInsertsPipeline()
- fault-remediation: Use GetPipelineBuilder().BuildQuarantinedAndDrainedNodesPipeline()
- node-drainer: Use GetPipelineBuilder().BuildNodeQuarantineStatusPipeline()

This is a pure refactoring commit - MongoDB behavior is unchanged.

Signed-off-by: Davanum Srinivas <dsrinivas@nvidia.com>

dims committed:
feat: Add owner reference to RebootNode CRs for automatic garbage collection

Add owner references from RebootNode CRs to the underlying Node objects.
This enables Kubernetes garbage collection to automatically clean up
RebootNode CRs when the corresponding Node is deleted.

Changes:
- Verify node exists before creating RebootNode CR
- Add owner reference pointing to the Node (with UID)
- Skip CR creation if node no longer exists
- Add nodeExistsFunc for testability

This is an intentional behavioral improvement:
- Prevents orphaned RebootNode CRs when nodes are deleted
- Eliminates need for manual CR cleanup
- Provides audit trail via owner reference

Signed-off-by: Davanum Srinivas <dsrinivas@nvidia.com>

dims committed:
feat: Add XID burst detector to health-events-analyzer

Add detection of XID error bursts to identify patterns of repeated XID
errors that may indicate a larger system issue requiring investigation.

Features:
- XID burst detector that tracks XID errors across time windows
- Configurable burst detection thresholds
- Integration with health-events-analyzer reconciler
- Parser updates for XID error pattern matching

This is a new feature unrelated to the PostgreSQL migration.

Signed-off-by: Davanum Srinivas <dsrinivas@nvidia.com>

dims committed:
chore: Add Bitnami PostgreSQL Helm chart as subchart

Vendor the Bitnami PostgreSQL Helm chart as a subchart for deploying
PostgreSQL alongside NVSentinel. This provides a production-ready
PostgreSQL deployment with:

- High availability support
- Built-in backup capabilities
- TLS/SSL support
- Resource management and limits
- Prometheus metrics integration

This is a vendored third-party chart requiring minimal review.

Signed-off-by: Davanum Srinivas <dsrinivas@nvidia.com>

dims committed:
feat: Add PostgreSQL client implementation in store-client

Implement PostgreSQL client that provides the same interface as the MongoDB
client, enabling seamless switching between database backends.

Features:
- PostgreSQLClient implementing DatabaseClient interface
- PostgreSQL change stream emulation using triggers and notifications
- PostgreSQL pipeline builder for change stream filtering
- Support for health event and maintenance event operations
- TLS/SSL support for secure connections
- Comprehensive test coverage

This enables NVSentinel to use PostgreSQL as an alternative to MongoDB.

Signed-off-by: Davanum Srinivas <dsrinivas@nvidia.com>

dims committed:
feat: Add PostgreSQL datastore provider

Implement the PostgreSQL datastore provider that integrates with the
datastore abstraction layer. This enables PostgreSQL as a first-class
database backend alongside MongoDB.

Features:
- PostgreSQL datastore implementing DataStore interface
- Health event store with full CRUD operations
- Maintenance event store for CSP events
- Change stream emulation using PostgreSQL LISTEN/NOTIFY
- Pipeline filter translation from MongoDB format to SQL
- SQL filter builder for database-agnostic query translation
- Comprehensive test coverage for all components

The provider is automatically registered when imported via the
providers package import.

Signed-off-by: Davanum Srinivas <dsrinivas@nvidia.com>

dims committed:
feat: Add PostgreSQL deployment support and update all modules

Complete the PostgreSQL integration with deployment configuration,
documentation, and module updates across all NVSentinel components.

Helm/Kubernetes:
- Add PostgreSQL values and TLS certificate templates
- Update all deployment templates with datastore environment variables
- Add datastore configmap for provider selection
- Add PostgreSQL-specific E2E test configuration

CI/CD:
- Add PostgreSQL E2E test jobs to workflows
- Add schema validation scripts

Documentation:
- Add PostgreSQL provider documentation
- Add PostgreSQL schema definition

Module updates:
- Update go.mod/go.sum across all modules for PostgreSQL dependencies
- Update fault-quarantine reconciler for datastore abstraction
- Update node-drainer for datastore integration
- Update event-exporter for PostgreSQL support
- Update test helpers for PostgreSQL compatibility

Signed-off-by: Davanum Srinivas <dsrinivas@nvidia.com>

lalitadithya commented on distros/kubernetes/nvsentinel/charts/csp-health-monitor/templates/deployment.yaml:
```diff
@@ -36,17 +36,54 @@ spec:
+      {{- if and .Values.global.datastore (eq .Values.global.datastore.provider "postgresql") }}
+      initContainers:
+        - name: fix-cert-permissions
+          image: "docker.io/bitnamilegacy/os-shell:12-debian-12-r30"
```
Can we make the image name a value? Some user may want to use their custom mirror images to avoid docker rate limits

lalitadithya commented on distros/kubernetes/nvsentinel/charts/event-exporter/templates/deployment.yaml:
```diff
@@ -36,10 +36,46 @@ spec:
+      {{- if and .Values.global.datastore (eq .Values.global.datastore.provider "postgresql") }}
+      initContainers:
+        - name: fix-cert-permissions
+          image: "docker.io/bitnamilegacy/os-shell:12-debian-12-r30"
```
Same comment as earlier

dims replied:
good call. will do.

lalitadithya commented on distros/kubernetes/nvsentinel/charts/fault-remediation/templates/deployment.yaml:
```diff
@@ -36,10 +36,43 @@ spec:
+      {{- if and .Values.global.datastore (eq .Values.global.datastore.provider "postgresql") }}
+      initContainers:
+        - name: fix-cert-permissions
+          image: "docker.io/bitnamilegacy/os-shell:12-debian-12-r30"
```
Same comment as earlier

dims replied:
good call. will do.

lalitadithya commented on distros/kubernetes/nvsentinel/charts/node-drainer/templates/deployment.yaml:
```diff
@@ -36,10 +36,43 @@ spec:
+      {{- if and .Values.global.datastore (eq .Values.global.datastore.provider "postgresql") }}
+      initContainers:
+        - name: fix-cert-permissions
+          image: "docker.io/bitnamilegacy/os-shell:12-debian-12-r30"
```
Same comment as earlier

lalitadithya commented on distros/kubernetes/nvsentinel/charts/postgresql/Chart.yaml:
```diff
@@ -0,0 +1,37 @@
+maintainers:
+- name: Broadcom, Inc. All Rights Reserved.
+  url: https://github.com/bitnami/charts
+name: postgresql
```
Considering that we recently move to using a operator for mongodb, should we consider the same here? Maybe something like https://github.com/CrunchyData/postgres-operator?

I'm also happy to stick with what we have now and iterate on it.

lalitadithya commented on distros/kubernetes/nvsentinel/templates/daemonset.yaml:
```diff
@@ -41,8 +41,42 @@ spec:
+      {{- if and .Values.global.datastore (eq .Values.global.datastore.provider "postgresql") }}
+      initContainers:
+        - name: fix-cert-permissions
+          image: "docker.io/bitnamilegacy/os-shell:12-debian-12-r30"
```
Same comment as earlier about using the image name from values

dims replied:
good call. will do.

lalitadithya commented on node-drainer/pkg/evaluator/evaluator.go:
```diff
@@ -268,14 +270,24 @@ func isNodeAlreadyDrained(ctx context.Context, database queue.DataStore, nodeNam
 	if !ok {
-		return false, fmt.Errorf("invalid healtheventstatus format for node %s", nodeName)
+		// Try camelCase (PostgreSQL JSON)
+		healthEventStatus, ok = document["healthEventStatus"].(map[string]interface{})
```
Would it be possible to handle this in the `result.Decode` so that we don't have to make this check in every call site?

lalitadithya commented on node-drainer/pkg/queue/queue.go:
```diff
@@ -56,18 +56,53 @@ func (m *eventQueueManager) EnqueueEventGeneric(ctx context.Context,
 
+// extractEventID extracts the document ID from an event map
+// Tries both MongoDB (_id) and PostgreSQL (id) formats
+func extractEventID(event datastore.Event) string {
```
Should we move this into the datastore package so that client don't have to be aware of the datatsore specific logic?

dims replied:
good call. switching to another method we have already in store-client.

lalitadithya commented on store-client/pkg/datastore/providers/postgresql/changestream.go:
```diff
@@ -0,0 +1,1555 @@
+	ModeNotify ChangeStreamMode = "notify"
+	// ModeHybrid uses LISTEN/NOTIFY with adaptive fallback polling (recommended)
+	ModeHybrid ChangeStreamMode = "hybrid"
+)
```
Just for my understanding, is the `notify` option experimental for us or is that an exterimental/alpha feature in Postgres?

I'm also wondering if we can simplify and we can pick only one mode.

lalitadithya commented on tests/helpers/kube.go:
```diff
@@ -172,6 +172,9 @@ func StartNodeLabelWatcher(ctx context.Context, t *testing.T, c klient.Client, n
 	// thead to read from the success channel. This is the desired behavior because we only want to have 1 UpdateFunc
 	// thread write true/false.
+	//
+	// NOTE: On ARM64 + PostgreSQL, rapid state transitions may occur faster than the Kubernetes watch
```
Curious about why it is an ARM64 + PostgreSQL and not just a PostgreSQL thing

dims committed:
incorporate feedback from review

Signed-off-by: Davanum Srinivas <dsrinivas@nvidia.com>

dims commented on store-client/pkg/datastore/providers/postgresql/database_client.go:
```diff
@@ -0,0 +1,1017 @@
+// NewPostgreSQLDatabaseClient creates a new PostgreSQL database client
+// Converts MongoDB-style PascalCase table names to PostgreSQL snake_case
+func NewPostgreSQLDatabaseClient(db *sql.DB, tableName string) client.DatabaseClient {
+	// Convert PascalCase to snake_case for PostgreSQL compatibility
```
I am afraid to touch the names of the tables and columns at this point. i can log this as a follow up for myself. is that ok?

dims commented on node-drainer/pkg/reconciler/reconciler.go:
```diff
@@ -226,6 +237,19 @@ func unmarshalGenericEvent(event map[string]interface{}, target interface{}) err
 }
 
+// extractEventID extracts the event ID from an event map (supports _id and id formats).
+func extractEventID(event datastore.Event) string {
```
good call. switching to another method we have already in store-client.

dims commented on node-drainer/main.go:
```diff
@@ -196,6 +197,13 @@ func startMetricsServer(g *errgroup.Group, gCtx context.Context, srv server.Serv
 func startEventWatcher(ctx context.Context, components *initializer.Components, criticalError chan<- error) {
 	go func() {
+		if components.EventWatcher == nil {
+			slog.Debug("No event watcher available")
```
totally. bumping up.

dims commented on health-events-analyzer/pkg/analyzer/xid_burst_detector.go:
i was ambivalent as well. i wanted to see how far i would get to replicate what we need minimally for running all the tests we have. Yes, we are dependent on mongodb syntax here heavily and we have to call it out as this is a critical component!

dims committed:
fix: use correct namespace for log-collector job operations

The DeleteAllLogCollectorJobs and WaitForLogCollectorJobStatus functions
were not specifying a namespace when listing/deleting jobs. This caused
the functions to operate on the wrong namespace (defaulting to the client
context namespace instead of nvsentinel where log-collector jobs run).

Fixed by using c.Resources(NVSentinelNamespace) for all job operations.

Signed-off-by: Davanum Srinivas <dsrinivas@nvidia.com>
